examples/chef/common/chef-fan-control-manager.cpp

```diff
@@ -99,98 +112,223 @@ Status ChefFanControlManager::HandleStep(StepDirectionEnum aDirection, bool aWra
-            // tracks the target for our product.
-            MatterReportingAttributeChangeCallback(mEndpoint, FanControl::Id, Attributes::PercentSetting::Id);
-            MatterReportingAttributeChangeCallback(mEndpoint, FanControl::Id, Attributes::PercentCurrent::Id);
+        // Determine if the speed change should also change the fan mode
+        if (mSpeedCurrent == 0)
+        {
+            FanControl::Attributes::FanMode::Set(mEndpoint, FanControl::FanModeEnum::kOff);
+        }
+        else if (mSpeedCurrent <= FAN_MODE_LOW_UPPER_BOUND)
+        {
+            FanControl::Attributes::FanMode::Set(mEndpoint, FanControl::FanModeEnum::kLow);
         }
+        else if (mSpeedCurrent <= FAN_MODE_MEDIUM_UPPER_BOUND)
+        {
+            FanControl::Attributes::FanMode::Set(mEndpoint, FanControl::FanModeEnum::kMedium);
+        }
+        else if (mSpeedCurrent <= FAN_MODE_HIGH_UPPER_BOUND)
+        {
+            FanControl::Attributes::FanMode::Set(mEndpoint, FanControl::FanModeEnum::kHigh);
+        }
+    }
+}
 
+void ChefFanControlManager::FanModeWriteCallback(FanControl::FanModeEnum aNewFanMode)
+{
+    ChipLogDetail(NotSpecified, "ChefFanControlManager::FanModeWriteCallback: %d", (uint8_t) aNewFanMode);
+    switch (aNewFanMode)
+    {
+    case FanControl::FanModeEnum::kOff: {
+        if (mSpeedCurrent != 0)
+        {
+            DataModel::Nullable<uint8_t> speedSetting(0);
+            SetSpeedSetting(speedSetting);
+        }
         break;
     }
-    default:
+    case FanControl::FanModeEnum::kLow: {
+        if (mSpeedCurrent < FAN_MODE_LOW_LOWER_BOUND || mSpeedCurrent > FAN_MODE_LOW_UPPER_BOUND)
+        {
+            DataModel::Nullable<uint8_t> speedSetting(FAN_MODE_LOW_LOWER_BOUND);
+            SetSpeedSetting(speedSetting);
+        }
         break;
     }
-
-    // Fall through goes to attribute store legacy handling.
-    return CHIP_NO_ERROR;
+    case FanControl::FanModeEnum::kMedium: {
+        if (mSpeedCurrent < FAN_MODE_MEDIUM_LOWER_BOUND || mSpeedCurrent > FAN_MODE_MEDIUM_UPPER_BOUND)
+        {
+            DataModel::Nullable<uint8_t> speedSetting(FAN_MODE_MEDIUM_LOWER_BOUND);
+            SetSpeedSetting(speedSetting);
+        }
+        break;
+    }
+    case FanControl::FanModeEnum::kOn:
+    case FanControl::FanModeEnum::kHigh: {
+        if (mSpeedCurrent < FAN_MODE_HIGH_LOWER_BOUND || mSpeedCurrent > FAN_MODE_HIGH_UPPER_BOUND)
+        {
+            DataModel::Nullable<uint8_t> speedSetting(FAN_MODE_HIGH_LOWER_BOUND);
+            SetSpeedSetting(speedSetting);
+        }
+        break;
+    }
+    case FanControl::FanModeEnum::kSmart:
+    case FanControl::FanModeEnum::kAuto: {
+        ChipLogProgress(NotSpecified, "ChefFanControlManager::FanModeWriteCallback: Auto");
+        break;
+    }
+    case FanControl::FanModeEnum::kUnknownEnumValue: {
+        ChipLogProgress(NotSpecified, "ChefFanControlManager::FanModeWriteCallback: Unknown");
```

Comment:
I'm new to this code but it looks like the cluster code only handles `kSmart` and `kOn`  on pre-callback. Everything else it just passes down: https://github.com/project-chip/connectedhomeip/blob/master/src/app/clusters/fan-control-server/fan-control-server.cpp#L183

Are you suggesting I go ahead and update the cluster code to check if it's `kUnknownValue` and return `Status::ConstraintError` instead of passing down?
I could do that. But searching the codebase, I don't see other clusters handling this on pre-callback, so I'm not sure what's the rationale here. Was the intention to just let the app/products handle this, maybe?

Reply:
> I'm new to this code but it looks like the cluster code only handles kSmart and kOn on pre-callback. 

That sounds like a bug in the cluster code; it should deal with invalid input in the way the spec specifies, so that applications do not have to.

>

Reply:
Makes sense. I'll address that in a separate PR as I don't want to mix changes to cluster code and example code together.

Reply:
Yes, separate PR is a good idea.